README.md
[<img src='https://s3.amazonaws.com/drivendata-public-assets/logo-white-blue.png' width='600'>](https://www.drivendata.org/)
<br><br>

<img src='https://s3.amazonaws.com/drivendata-public-assets/swe-reclamation1.jpg' width='600'>

# Snowcast Showdown

## Goal of the Competition

Seasonal mountain snowpack is a critical water resource throughout the Western U.S. Snowpack acts as a natural reservoir by storing precipitation throughout the winter months and releasing it as snowmelt when temperatures rise during the spring and summer. This meltwater becomes runoff and serves as a primary freshwater source for major streams, rivers and reservoirs. As a result, snowpack accumulation on high-elevation mountains significantly influences streamflow as well as water storage and allocation for millions of people.

**The goal of the Snowcast Showdown challenge was to estimate snow water equivalent (SWE) in real-time each week for 1km x 1km grid cells across the Western U.S.** Getting better high-resolution SWE estimates for mountain watersheds and headwater catchments will help to improve runoff and water supply forecasts, which in turn will help reservoir operators manage limited water supplies. Improved SWE information will also help water managers respond to extreme weather events such as floods and droughts.

## What's in this Repository

This repository contains code from winning competitors in the [Snowcast Showdown](https://www.drivendata.org/competitions/90/competition-reclamation-snow-water-eval/) DrivenData challenge.

**Winning code for other DrivenData competitions is available in the [competition-winners repository](https://github.com/drivendataorg/competition-winners).**

## Winning Submissions

### Real-time Prediction

Place |Team or User | Score | Summary of Model
--- | --- | --- | ---
1   | FBykov | 3.8784 | Two layers perceptrons are used to estimate the mean, the spread, and the mapping in latent space. The final result is a mix of 63 inhomogeneous kriging models. The models have 8 different configurations and use different train-validation splitting.
2   | UltimateHydrology | 3.9746 | The solution is based on different SOTA implementations of Gradient Boosting Machine algorithm: XGBoost, LightGBM, and CatBoost, and their ensembles. The top features of 121 features set included ground snow measure data (SNOTEL, CDEC), and remote sensing of snow cover (MODIS Terra MOD10A1). The top 1-4 features include physically-based indirect predictors of SWE: seasonal cumulative sum of solid precipitation, seasonal average values of air temperature and the mean seasonal value of solar radiation.
3   | TeamUArizona | 3.9900 | Multilinear Regression (MLR) models are used to predict SWE based on the provided snow station data. Models are trained to predict either ground measurements of SWE (if there are enough measurements for a particular grid cell), or SWE data taken from a gridded SWE dataset called the University of Arizona (UA) SWE dataset (if they are not). In addition to the MLR models, there is also code to fill missing snow station predictor data and to perform some bias correction of the models if necessary. Ultimately, model predictions are made by averaging an ensemble of MLR models.
4   | leigh.plt | 4.0892 | We used a neural network model with different layer architectures: Fourier neural operator (FNO), convolution layers, embeddings, and linear transformation. This architecture allows us to combine data of different nature: images, historical data and classification labels.
5   | oshbocker | 4.1828 | I used an ensemble of three different gradient boosting implementations, LightGBM, XGBoost, and Catboost. Since gradient boosting naturally works with tabular data, I used the mean and in some cases the variance of pixel values from the Modis data and the DEM over an entire grid cell. I also created a feature representing southern sun exposure and a feature capturing the number of days since the beginning of the snow season. The way I incorporated the ground measures was to use the relative SWE, compared to historical average for a ground measure station. Then for each grid cell I took the 15 nearest neighbors relative SWE.

Additional solution details can be found in the `reports` folder inside the directory for each submission.

### Modeling Report

Everyone who successfully submitted a model for real-time evaluation was invited to submit a [modeling report](https://www.drivendata.org/competitions/86/competition-reclamation-snow-water-dev/page/415/#report) that discusses their solution methodology and explains its performance on historical data. Reports were evaluated on the following criteria:

- **Interpretability**: To what extent can a person understand the outcome of the solution methodology given its inputs?
- **Robustness**: Do solutions provide high performance over a broad range of geographic and climate conditions?
- **Rigor**: To what extent is the report built on sound, sophisticated quantitative analysis and a performant statistical model?
- **Clarity**: How clearly are underlying patterns exposed, communicated, and visualized?

Place | Team or User | Place in Prediction Competition | Link
--- | --- | --- | ---
1   | oshbocker | 5th | [Read the report](https://github.com/drivendataorg/snowcast-showdown/blob/main/5th%20Place/reports/Model%20Report.pdf)
2   | UltimateHydrology |  2nd | [Read the report](https://github.com/drivendataorg/snowcast-showdown/blob/main/2nd%20Place/reports/Model%20report%20UltimateHydrology.pdf)
3   | TeamUArizona | 3rd| [Read the report](https://github.com/drivendataorg/snowcast-showdown/blob/main/3rd%20Place/reports/TeamUArizona_Modeling_Report.pdf)

### Regional Prizes

[Regional prizes](https://www.drivendata.org/competitions/90/competition-reclamation-snow-water-eval/page/533/#regional) are awarded for regional performance in the Sierras and Central Rockies. Submissions were evaluated against regional grid cells in the real-time evaluation period. Final rankings were determined by the scoring metric, RMSE.

#### Sierras Regional Prize

Place | Team or User | Place in Prediction Competition
--- | --- | ---
1   | leigh.plt | 4th
2   | FBykov | 1st
3   | UltimateHydrology | 2nd

#### Central Rockies Regional Prize

Place | Team or User | Place in Prediction Competition
--- | --- | ---
1   | UltimateHydrology | 2nd
2   | oshbocker | 5th
3   | leigh.plt | 4th

1st Place/CHANGELOG.md
2022-02-10 - 2022-03-10
Initial submission

2022-03-17
Added visualization:
Changed scripts in the directory src/visualization
In main.py added lines 152-159
        figdir = path.join (maindir, 'reports', 'figures')
        makedirs (figdir,exist_ok=True)
        results = models.applymodels (inputs['test'], modelslist, average=False).clamp_(min=0.)
        # monests = visualization.monthests (inputs, dates, uregions, result)
        visualization.temporal(inputs['test'], dates['test'], results, result, uregions, figdir)
        visualization.spatial(inputs['test'], dates['test'], result, figdir)
        visualization.importance(inputs['test'], dates['test'], modelslist, result, arglist, nousest, figdir)
        visualization.latentsd(inputs['test'], days['test'], 8, modelslist[1], figdir)

2022-03-24 - 2022-06-30
Improved logging
In src/main.py added line 13
from traceback import format_exc
In src/main.py changed lines 131-154 (after writing submission file)
	print ('Mean values: ' + str({key: np.round(sub[key].mean(),4) for key in sub}))
        try:
            isubs = pd.DataFrame({'week': [iweek], 'day': [evalday], 'file': [file]})
            subslistfile = path.join(workdir,'submissionslist.csv')            
            if path.isfile(subslistfile):
                subslist = pd.read_csv(subslistfile)
                pd.concat((pd.read_csv(subslistfile).set_index(subslist.columns[0]), isubs)).to_csv(subslistfile)
                isst = (inputs['test']['yemb'][0]>0).cpu().numpy()
                for file in subslist['file']:
                    try:
                        old = pd.read_csv(path.join(workdir,file)).set_index('cell_id')
                        print (f'Compare with {file}\n'+
                               str({key: [np.round((sub[key]-old[key])[isst].mean(),4), np.round(np.abs(sub[key]-old[key])[isst].mean(),4),
                                          np.round((sub[key]-old[key])[~isst].mean(),4), np.round(np.abs(sub[key]-old[key])[~isst].mean(),4)]
                                for key in old if np.isfinite(sub[key]).sum()>0}))
                    except:
                        print(format_exc())
            else:
                isubs.to_csv(subslistfile)
            from shutil import copyfile
            copyfile(path.join(workdir,'ground_measures_features.csv'), path.join(subdir,'ground_measures_features.csv'))
            copyfile(logfile, path.join(subdir,datestr+'.log'))
        except:
            print(format_exc())
In src/features/inputs.py added lines 28-29
	stswe_test = pd.read_csv(file)
        file = path.join(workdir,'ground_measures_features.csv')
        stswe_test.to_csv(file)

3rd Place/RunModels.py
#******************************************************************************
#  Copyright (c) 2022, Patrick Broxton <broxtopd@arizona.edu>
# 
#  Permission is hereby granted, free of charge, to any person obtaining a
#  copy of this software and associated documentation files (the "Software"),
#  to deal in the Software without restriction, including without limitation
#  the rights to use, copy, modify, merge, publish, distribute, sublicense,
#  and/or sell copies of the Software, and to permit persons to whom the
#  Software is furnished to do so, subject to the following conditions:
# 
#  The above copyright notice and this permission notice shall be included
#  in all copies or substantial portions of the Software.
# 
#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
#  OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
#  THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
#  DEALINGS IN THE SOFTWARE.
#******************************************************************************

import sys
import os
from datetime import datetime
import urllib.request
import io
import numpy as np
import geojson
import pickle
import time

tic = time.perf_counter()

Identifiers = sys.argv[1]
SubmissionID = sys.argv[2]
if ',' in Identifiers:
    ModelIdentifiers = Identifiers.split(',')
else:
    ModelIdentifiers = [Identifiers]

#### Get Data ####
print('Getting Data')

# Download the updated ground measures data
with urllib.request.urlopen('https://drivendata-public-assets.s3.amazonaws.com/ground_measures_features.csv') as response:
   html = response.read().decode("utf-8")
buf = io.StringIO(html)
num_lines = sum(1 for line in buf)  # Number of lines in file

# Read the submission dates
buf = io.StringIO(html)
tline = buf.readline()
header = tline.split(',')
Dates_submit_all = []
for element in header:
    if element[0:2] == '20':
        Dates_submit_all.append(datetime.strptime(element.strip(), '%Y-%m-%d'))

# Only use those in the past
num_times = 0
for Date in Dates_submit_all:
    if Date < datetime.today():
        num_times = num_times + 1

# Read the data
tline = buf.readline()
i = 0
GroundSWE = np.empty([num_lines-1,num_times])
GroundSWE[:] = np.nan
while not (tline == ''):
    fields = tline.split(',')
    for d in range(num_times):
## 2022-06-30 PDB: For the 6/30 submission, the previous code didn't correctly read the data
        # if not (fields[d+1] == ''):
        if not (fields[d+1] == '') and len(fields[d+1]) > 1:
            GroundSWE[i,d] = float(fields[d+1])
    i = i+1
    tline = buf.readline()


# Get IDs of all features
with open('Data/grid_cells.geojson') as f:
    gj = geojson.load(f)
features = gj['features']
cell_ids = []
for feature in features:
    cell_ids.append(feature['properties']['cell_id'])

# Define variables for later
NSites_all = len(cell_ids)
NDates = len(GroundSWE[0, :])
NSites_ground = len(GroundSWE[:, 1])
Dates_submit = Dates_submit_all[0:NDates]

#### Run ML Models ####

print('Running MLR Models')
OutputSWE = np.empty([NSites_all, NDates, len(ModelIdentifiers)])
e = 0

# Loop through the given model identifiers in the ensemble
for ModelIdentifier in ModelIdentifiers:

    # Load the Model Weights
    file = open(r'Models/' + ModelIdentifier + '.dat', 'rb')
    md = pickle.load(file)
    file.close()
    GroundModels = md['GroundModels']
    Models = md['Models']

    # Fill in missing ground measures data
    GroundSWE_filled = np.empty(GroundSWE.shape)
    GroundSWE_filled[:] = 0
    for i in range(NSites_ground):
        a = GroundModels[i]['a']
        weights = GroundModels[i]['weights']
        locs = GroundModels[i]['locs']
        bias = GroundModels[i]['bias']
        if len(GroundModels[i]['a']) > 0:
            for d in range(NDates):
                if not np.nansum(weights * ~np.isnan(GroundSWE[locs, d])) == 0:
                    GroundSWE_filled[i, d] = np.nansum((a * GroundSWE[locs, d]) * weights) / np.nansum(
                        weights * ~np.isnan(GroundSWE[locs, d])) * bias
                else:
                    GroundSWE_filled[i, d] = 0
        else:
            GroundSWE_filled[i, :] = np.nan

    # Use the historical observed values to better constrain filled values
    relax_factor = 0.05
    bias_prev = np.empty(NSites_ground)
    bias_prev[:] = 1
    for d in range(NDates):
        bias = GroundSWE[:, d] / (GroundSWE_filled[:, d] + 1E-5)
        bias[GroundSWE_filled[:, d] == 0] = np.nan
        bias[bias > 5] = 5
        nanlocs = np.isnan(bias)
        bias[nanlocs] = bias_prev[nanlocs]
        GroundSWE_filled[:, d] = GroundSWE_filled[:, d] * bias
        bias_prev = (bias + relax_factor) / (1 + relax_factor)

    # Run the MLR model for each location
    AllSWE_ML = np.empty([NSites_all, NDates])
    AllSWE_ML[:] = 0
    for i in range(NSites_all):
        # Do not run if model doesn't exist
        if not (Models[i]['mdl'] == []):
            x = np.transpose(GroundSWE_filled[Models[i]['locs'], :])
            y = Models[i]['mdl'].predict(x)
            bias = Models[i]['bias']
            y = np.transpose(y) + bias['c']
            y[y < 0] = 0
            AllSWE_ML[i, :] = bias['a'] * y ** bias['b']

            # Boot jumps that are outside the allowable range (3x the 1st and 99th percentiles of observed SWE changes)
            for c in range(NDates):
                if c > 0:
                    diff = AllSWE_ML[i, c] - AllSWE_ML[i, c - 1]
                    if diff < Models[i]['minmax'][0] or diff > Models[i]['minmax'][1]:
                        AllSWE_ML[i, c] = AllSWE_ML[i, c - 1]

    OutputSWE[:, :, e] = AllSWE_ML
    e = e + 1

OutputSWE = np.mean(OutputSWE, axis=2)

#### Generate the Submission File ####
print('Generating the submission file')

# Create output directory if it does not exist
if not os.path.exists('Output CSV'):
    os.makedirs('Output CSV')

# Write the output file
f = open('Output CSV/' + SubmissionID + '.csv', 'w')
f.write(',2022-01-13,2022-01-20,2022-01-27,2022-02-03,2022-02-10,2022-02-17,2022-02-24,2022-03-03,2022-03-10,2022-03-17,2022-03-24,2022-03-31,2022-04-07,2022-04-14,2022-04-21,2022-04-28,2022-05-05,2022-05-12,2022-05-19,2022-05-26,2022-06-02,2022-06-09,2022-06-16,2022-06-23,2022-06-30\n')
for i in range(NSites_all):
    f.write(cell_ids[i])
    for d in range(NDates):
        f.write(',{:.2f}'.format(OutputSWE[i, d]))
    f.write('\n')
f.close()

toc = time.perf_counter()
print(f"Process Done. Elapsed time is {toc - tic:0.2f} seconds")
